examples/pytorch/lda/lda_model.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2021 Yifei Ma
+# with references from "sklearn.decomposition.LatentDirichletAllocation"
+# with the following original authors:
+# * Chyi-Kwei Yau (the said scikit-learn implementation)
+# * Matthew D. Hoffman (original onlineldavb implementation)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import os, functools, warnings
+import numpy as np, scipy as sp
+import torch
+import dgl
+from dgl import function as fn
+
+
+def _lbeta(alpha, axis):
+    return torch.lgamma(alpha).sum(axis) - torch.lgamma(alpha.sum(axis))
+
+
+_sklearn_random_init = torch.distributions.gamma.Gamma(100, 0.01)
+
+
+def _edge_update(edges, step_size=1):
+    """ the Gibbs posterior distribution of z propto theta*beta.
+    As step_size -> infty, the result becomes MAP estimate on the dst nodes.
+    """
+    q = edges.src['weight'] * edges.dst['weight']
+    marg = q.sum(axis=1, keepdims=True) + np.finfo(float).eps
+    p = q / marg
+
+    return {
+        'z': p * step_size,
+        'edge_elbo': marg.squeeze(1).log() * step_size,
+    }
+
+
+def _weight_exp(z, ntype, prior):
+    """Node weight is approximately normalized for VB along the ntype
+    direction.
+    """
+    prior = prior + z * 0 # convert numpy to torch
+    gamma = prior + z
+
+    axis = 1 if ntype == 'doc' else 0 # word
+    Elog = torch.digamma(gamma) - torch.digamma(gamma.sum(axis, keepdims=True))
+
+    return Elog.exp()
+
+
+def _node_update(nodes, prior):
+    return {
+        'z': nodes.data['z'],
+        'weight': _weight_exp(nodes.data['z'], nodes.ntype, prior)
+    }
+
+
+def _update_all(G, ntype, prior, step_size, return_obj=False):
+    """Follows Eq (5) of Hoffman et al., 2010.
+    """
+    G_prop = G.reverse() if ntype == 'doc' else G # word
+    msg_fn = lambda edges: _edge_update(edges, step_size)
+    node_fn = lambda nodes: _node_update(nodes, prior)
+
+    G_prop.update_all(msg_fn, fn.sum('z','z'), node_fn)
+
+    if return_obj:
+        G_prop.update_all(msg_fn, fn.sum('edge_elbo', 'edge_elbo'))
+
+    G.nodes[ntype].data.update(G_prop.nodes[ntype].data)
+    return dict(G.nodes[ntype].data)
+
+
+class LatentDirichletAllocation:
+    """LDA model that wors with a HeteroGraph with doc/word node types.
```

Comment:
```suggestion
    """LDA model that works with a HeteroGraph with doc/word node types.
```

Reply:
Also, I assume that the node type and edge type have to be "doc", "topic", and "word"?

Reply:
No requirements in edge type; only node types.